[SLP] Make getSameOpcode support interchangeable instructions.

HanKuanChen opened the pull request:
We use the term "interchangeable instructions" to refer to different operators that have the same meaning (e.g., `add x, 0` is equivalent to `mul x, 1`).
Non-constant values are not supported, as they may incur high costs with little benefit.

HanKuanChen committed:
[SLP] Pre-commit test.

HanKuanChen committed:
[SLP] Make getSameOpcode support different instructions if they have
same semantics.

HanKuanChen committed:
[SLP] Fix isOpcodeOrAlt cannot find interchangeable instruction.

HanKuanChen committed:
support Or

HanKuanChen committed:
[SLP] NFC. Update test.

HanKuanChen committed:
support Xor

HanKuanChen committed:
fix undef deprecator issue

HanKuanChen committed:
clang-format

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,207 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+      return nullptr;
+    if (auto *CI = dyn_cast<ConstantInt>(C)) {
+      return CI;
+    } else if (auto *CDV = dyn_cast<ConstantDataVector>(C)) {
+      if (auto *CI = dyn_cast_if_present<ConstantInt>(CDV->getSplatValue()))
+        return CI;
+    }
```
```suggestion
    if (auto *CDV = dyn_cast<ConstantDataVector>(C)) {
      if (auto *CI = dyn_cast_if_present<ConstantInt>(CDV->getSplatValue()))
        return CI;
    }
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,207 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+
+public:
+  InterchangeableInstruction(Instruction *MainOp) : MainOp(MainOp) {}
+  virtual bool isSame(Instruction *I) {
```
We'd better to avoid virtual where possible and replace by static dispatching, if possible

alexey-bataev replied:
How about this one?

HanKuanChen replied:
Probably cannot. I expected to add other class here. For example,
```
define void @test(<16 x i8> %81, ptr %output) {
entry:
  %shuffle.i = shufflevector <16 x i8> %81, <16 x i8> poison, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  %shuffle.i109 = shufflevector <16 x i8> %81, <16 x i8> poison, <8 x i32> <i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15>
  %vmovl.i108 = zext <8 x i8> %shuffle.i to <8 x i16>
  %vmovl.i = zext <8 x i8> %shuffle.i109 to <8 x i16>
  %vext = shufflevector <8 x i16> %vmovl.i108, <8 x i16> %vmovl.i, <8 x i32> <i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8>
  %vext20 = shufflevector <8 x i16> %vmovl.i108, <8 x i16> %vmovl.i, <8 x i32> <i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9>
  %vext29 = shufflevector <8 x i16> %vmovl.i108, <8 x i16> %vmovl.i, <8 x i32> <i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10>
  %gep0 = getelementptr inbounds i16, ptr %output, i64 0
  %gep1 = getelementptr inbounds i16, ptr %output, i64 8
  %gep2 = getelementptr inbounds i16, ptr %output, i64 16
  %gep3 = getelementptr inbounds i16, ptr %output, i64 24
  store <8 x i16> %vext, ptr %gep0, align 4
  store <8 x i16> %vmovl.i108, ptr %gep1, align 4
  store <8 x i16> %vext20, ptr %gep2, align 4
  store <8 x i16> %vext29, ptr %gep3, align 4
  ret void
}
```
zext and shufflevector are interchangeable. `%vmovl.i108` can be `shufflevector <8 x i16> %vmovl.i108, <8 x i16> %vmovl.i, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>`.
```
  %vmovl.i108 = shufflevector <8 x i16> %vmovl.i108, <8 x i16> %vmovl.i, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  %vext = shufflevector <8 x i16> %vmovl.i108, <8 x i16> %vmovl.i, <8 x i32> <i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8>
  %vext20 = shufflevector <8 x i16> %vmovl.i108, <8 x i16> %vmovl.i, <8 x i32> <i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9>
  %vext29 = shufflevector <8 x i16> %vmovl.i108, <8 x i16> %vmovl.i, <8 x i32> <i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10>
```
virtual can help me do that.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,207 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+  using InterchangeableInstruction::InterchangeableInstruction;
+  bool isSame(Instruction *I) override {
+    unsigned Opcode = I->getOpcode();
+    if (!is_contained(SupportedOp, Opcode))
+      return false;
```
Better to make it a switch to avoid linear complexity here?

HanKuanChen replied:
SupportedOp is sorted. We can use binary search here.

HanKuanChen replied:
binary_search is used. see https://github.com/llvm/llvm-project/pull/127450/commits/f422a5942931445fba3dd949521711457c3d9307

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,207 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+          return true;
+      }
+    }
+    if (is_contained(CandidateOp, Opcode)) {
+      CandidateOp = {Opcode};
+      return true;
+    }
```
Same, make it switch somehow (in a separate function?)?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,207 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+  }
+  SmallVector<Value *>
+  getInterchangeableInstructionOps(Instruction *I) override {
+    assert(is_contained(SupportedOp, I->getOpcode()));
```
Add a message

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1085,6 +1294,29 @@ static InstructionsState getSameOpcode(ArrayRef<Value *> VL,
+            for (Value *V : VL) {
+              if (isa<PoisonValue>(V))
+                continue;
+              if (cast<Instruction>(V)->getOpcode() ==
+                  InterchangeableInstructionOpcode)
+                return cast<Instruction>(V);
```
Avoid double cast

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2447,42 +2679,44 @@ class BoUpSLP {
       OpsVec.resize(NumOperands);
       unsigned NumLanes = VL.size();
-      for (unsigned OpIdx = 0; OpIdx != NumOperands; ++OpIdx) {
+      for (unsigned OpIdx = 0; OpIdx != NumOperands; ++OpIdx)
```
```suggestion
      for (unsigned OpIdx : seq<unsigned>(NumOperands))
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2447,42 +2679,44 @@ class BoUpSLP {
-          if (isa<PoisonValue>(VL[Lane])) {
-            if (auto *EI = dyn_cast<ExtractElementInst>(MainOp)) {
-              if (OpIdx == 0) {
+      for (unsigned Lane = 0; Lane != NumLanes; ++Lane) {
```
```suggestion
      for (unsigned Lane : seq<unsigned>(NumLanes)) {
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2447,42 +2679,44 @@ class BoUpSLP {
-              }
-            } else if (auto *EV = dyn_cast<ExtractValueInst>(MainOp)) {
-              if (OpIdx == 0) {
+              } else if (auto *EV = dyn_cast<ExtractValueInst>(MainOp)) {
```
```suggestion
              if (auto *EV = dyn_cast<ExtractValueInst>(MainOp)) {
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -8501,8 +8735,12 @@ void BoUpSLP::buildTree_rec(ArrayRef<Value *> VL, unsigned Depth,
 
+  SmallVector<Value *> MainOpIsTheFirst(UniqueValues);
+  auto MainOpIter = find(MainOpIsTheFirst, S.getMainOp());
+  std::rotate(MainOpIsTheFirst.begin(), MainOpIter, std::next(MainOpIter));
+
   std::optional<ScheduleData *> Bundle =
-      BS.tryScheduleBundle(UniqueValues, this, S);
+      BS.tryScheduleBundle(MainOpIsTheFirst, this, S);
```
Why do you need this?

HanKuanChen replied:
A Bundle is built in the order of VL.
```
BoUpSLP::ScheduleData *
BoUpSLP::BlockScheduling::buildBundle(ArrayRef<Value *> VL) {
  ScheduleData *Bundle = nullptr;
  ScheduleData *PrevInBundle = nullptr;
  for (Value *V : VL) {
    if (doesNotNeedToBeScheduled(V))
      continue;
    ScheduleData *BundleMember = getScheduleData(V);
    assert(BundleMember &&
           "no ScheduleData for bundle member "
           "(maybe not in same basic block)");
    assert(BundleMember->isSchedulingEntity() &&
           "bundle member already part of other bundle");
    if (PrevInBundle) {
      PrevInBundle->NextInBundle = BundleMember;
    } else {
      Bundle = BundleMember;
    }
```
That means the first one Value which is NOT `doesNotNeedToBeScheduled` will be the `FirstInBundle`.
When SLP calls `cancelScheduling`, `InstructionsState.MainOp` will be passed into `cancelScheduling`.
```
  auto *Bundle = buildBundle(VL);
  TryScheduleBundleImpl(ReSchedule, Bundle);
  if (!Bundle->isReady()) {
    cancelScheduling(VL, S.getMainOp());
```
The `Bundle` is from `getScheduleData(OpValue)`. But the latter code assume it is the first one in a bundle (by `Bundle->isSchedulingEntity()`).
```
void BoUpSLP::BlockScheduling::cancelScheduling(ArrayRef<Value *> VL,
                                                Value *OpValue) {
  if (isa<PHINode>(OpValue) || isVectorLikeInstWithConstOps(OpValue) ||
      doesNotNeedToSchedule(VL))
    return;

  if (doesNotNeedToBeScheduled(OpValue))
    OpValue = *find_if_not(VL, doesNotNeedToBeScheduled);
  ScheduleData *Bundle = getScheduleData(OpValue);
  LLVM_DEBUG(dbgs() << "SLP:  cancel scheduling of " << *Bundle << "\n");
  assert(!Bundle->IsScheduled &&
         "Can't cancel bundle which is already scheduled");
  assert(Bundle->isSchedulingEntity() &&
         (Bundle->isPartOfBundle() || needToScheduleSingleInstruction(VL)) &&
         "tried to unbundle something which is not a bundle");
```
This is not true for interchangeable instructions. For example,
```
define i32 @test() {
entry:
  %mul = mul i32 1, 0
  %shl = shl i32 %mul, 0
  %xor1 = xor i32 %shl, %mul
  ret i32 %xor1
}
```
VL is
```
%shl = shl i32 %mul, 0
%mul = mul i32 1, 0
```
The MainOp is `%mul` here.

HanKuanChen replied:
`assert(VL[0] == S.getMainOp() && "MainOp must be the first element of VL.");` is added in the beginning of `tryScheduleBundle`.

alexey-bataev replied:
Better to adjust cancelScheduling, the first item in the bundle is just a Bundle->FirstInBundle

HanKuanChen replied:
see https://github.com/llvm/llvm-project/pull/127450/commits/968f3463f6833675e504b8fb159224cd6176089f

alexey-bataev commented on llvm/test/Transforms/SLPVectorizer/AArch64/gather-with-minbith-user.ll:
```diff
@@ -5,7 +5,9 @@ define void @h() {
 ; CHECK-NEXT:  entry:
 ; CHECK-NEXT:    [[ARRAYIDX2:%.*]] = getelementptr i8, ptr null, i64 16
-; CHECK-NEXT:    store <8 x i16> zeroinitializer, ptr [[ARRAYIDX2]], align 2
+; CHECK-NEXT:    [[ARRAYIDX18:%.*]] = getelementptr i8, ptr null, i64 24
+; CHECK-NEXT:    store <4 x i16> zeroinitializer, ptr [[ARRAYIDX2]], align 2
+; CHECK-NEXT:    store <4 x i16> zeroinitializer, ptr [[ARRAYIDX18]], align 2
```
Regression?

HanKuanChen replied:
see https://github.com/llvm/llvm-project/pull/127450/commits/29c8cff49c65814c589f0f8499a22d093ea46dc4

alexey-bataev commented on llvm/test/Transforms/SLPVectorizer/X86/non-power-2-num-elems-reused.ll:
```diff
@@ -5,11 +5,19 @@ define i64 @test() {
+; CHECK-NEXT:    [[TMP0:%.*]] = call <4 x i64> @llvm.vector.extract.v4i64.v8i64(<8 x i64> zeroinitializer, i64 0)
+; CHECK-NEXT:    [[RDX_OP:%.*]] = or <4 x i64> [[TMP0]], zeroinitializer
+; CHECK-NEXT:    [[TMP1:%.*]] = call <8 x i64> @llvm.vector.insert.v8i64.v4i64(<8 x i64> zeroinitializer, <4 x i64> [[RDX_OP]], i64 0)
+; CHECK-NEXT:    [[OP_RDX:%.*]] = call i64 @llvm.vector.reduce.or.v8i64(<8 x i64> [[TMP1]])
```
Regression?

alexey-bataev replied:
Would be good somehow to estimate here, which one is better, and select best solution

HanKuanChen replied:
before this PR, SLP are trying to vectorize VL with size 12.
```%xor148.2.i = xor i32 0, 0
%xor148.2.i.1 = xor i32 0, 0
%xor148.2.i.2 = xor i32 0, 0
%xor148.2.i.3 = xor i32 0, 0
%xor148.2.i.4 = xor i32 0, 0
%xor148.2.i.5 = xor i32 0, 0
%xor148.2.i.6 = xor i32 0, 0
%xor148.2.i.7 = xor i32 0, 0
%or54.i.i.6 = or i32 %xor148.2.i.6, 0
i32 poison
i32 poison
i32 poison
```
the code has a check like this
```
      if (VL.size() <= 2 || LoadEntriesToVectorize.contains(Idx) ||
          !(!E.hasState() || E.getOpcode() == Instruction::Load ||
            E.isAltShuffle() || !allSameBlock(VL)) ||
          allConstant(VL) || isSplat(VL))
        continue;
```
VL is alternate shuffle (xor and or) and SLP will try to combine from VL[0] to VL[7]. (`CombinedEntriesWithIndices`)
VL is NOT alternate shuffle when this PR is applied.

alexey-bataev replied:
Need to fix this check, if it causes the regression

HanKuanChen replied:
But it should be in another PR. Even without this PR, SLP can still get the same result if `%or54.i.i.6` uses xor. It is not related to interchangeable instruction but SLP issue.

alexey-bataev commented on llvm/test/Transforms/SLPVectorizer/isOpcodeOrAlt.ll:
```diff
@@ -0,0 +1,38 @@
+; CHECK-NEXT:    [[GEP0:%.*]] = getelementptr inbounds i32, ptr [[A:%.*]], i64 0
+; CHECK-NEXT:    [[GEP4:%.*]] = getelementptr inbounds i32, ptr [[B:%.*]], i64 0
+; CHECK-NEXT:    [[TMP0:%.*]] = load <4 x i32>, ptr [[GEP0]], align 4
+; CHECK-NEXT:    [[TMP1:%.*]] = shl <4 x i32> [[TMP0]], <i32 1, i32 0, i32 1, i32 0>
+; CHECK-NEXT:    [[TMP2:%.*]] = mul <4 x i32> [[TMP0]], <i32 1, i32 0, i32 1, i32 0>
+; CHECK-NEXT:    [[TMP3:%.*]] = shufflevector <4 x i32> [[TMP1]], <4 x i32> [[TMP2]], <4 x i32> <i32 0, i32 1, i32 6, i32 3>
```
Shall we avoid alternate vectorization here too?

HanKuanChen replied:
see https://github.com/llvm/llvm-project/pull/127450/commits/0e8d5670dca3701467b9116bfe5cb3e49af61ac0

alexey-bataev commented on llvm/test/Transforms/SLPVectorizer/shuffle-mask-resized.ll:
```diff
@@ -12,9 +12,7 @@ define i32 @test() {
-; CHECK-NEXT:    [[TMP3:%.*]] = add <2 x i32> zeroinitializer, [[TMP2]]
-; CHECK-NEXT:    [[TMP4:%.*]] = or <2 x i32> zeroinitializer, [[TMP2]]
-; CHECK-NEXT:    [[TMP5]] = shufflevector <2 x i32> [[TMP3]], <2 x i32> [[TMP4]], <2 x i32> <i32 0, i32 3>
+; CHECK-NEXT:    [[TMP5]] = or <2 x i32> zeroinitializer, [[TMP2]]
```
I suppose it should be `add`, not `or`. I suppose, more relaxed version of the opcode must always be translated to some more `strict` ops, like `or`->`add`, `add`->`shl`, `shl`->`mul`, etc.

HanKuanChen replied:
The problem is 0 is the left hand side of `%or = or i32 0, %phi`.

HanKuanChen committed:
apply comment

HanKuanChen committed:
Merge BinOpIsNoOp and MulAndShlWithConstantInt to InterchangeableBinOp.

HanKuanChen committed:
add SeenBefore

HanKuanChen committed:
make isBinOpWithConstantInt support left hand side operand

HanKuanChen committed:
prefer AltOp instead of interchangeable instructions of MainOp

HanKuanChen committed:
rename getInterchangeableInstructionOpcode to getOpcode and getInterchangeableInstructionOps to getOperand

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,267 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+
+public:
+  InterchangeableInstruction(Instruction *MainOp) : MainOp(MainOp) {}
+  virtual bool isSame(Instruction *I) {
```
const?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,267 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+/// - Get the operands for the interchangeable form (getOperand)
+class InterchangeableInstruction {
+protected:
+  Instruction *const MainOp;
```
```suggestion
  Instruction *const MainOp = nullptr;
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,267 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+      case Instruction::Shl:
+        if (CIValue.isZero())
+          return true;
+        return tryAnd(0b101);
```
Avoid using explicit magic numbers

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,267 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+      ToCIValue = APInt::getZero(FromCIValueBitWidth);
+      break;
+    }
+    auto LHS = MainOp->getOperand(1 - Pos);
+    auto RHS = ConstantInt::get(MainOp->getOperand(Pos)->getType(), ToCIValue);
```
Expand `auto` here

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,267 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+    if (C->isSame(I) && C->isSame(MainOp))
+      return std::make_pair(MainOp, C->getOperand(MainOp));
+  Candidate = getInterchangeableInstruction(I);
+  for (std::unique_ptr<InterchangeableInstruction> &C : Candidate)
+    if (C->isSame(I) && C->isSame(AltOp))
+      return std::make_pair(AltOp, C->getOperand(AltOp));
```
Why do you need this lookup and cannot return it as a result of `getInterchangeableInstruction`?

HanKuanChen replied:
follow https://github.com/llvm/llvm-project/pull/127450#discussion_r1965026667

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1085,6 +1354,28 @@ static InstructionsState getSameOpcode(ArrayRef<Value *> VL,
+            for (Value *V : VL) {
+              if (isa<PoisonValue>(V))
+                continue;
+              Instruction *Inst = cast<Instruction>(V);
```
```suggestion
              auto *Inst = cast<Instruction>(V);
```

HanKuanChen committed:
MainOp may not be the first insturction in a bundle

HanKuanChen committed:
apply comment

HanKuanChen commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,267 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+
+  /// Return a non-nullptr if either operand of I is a ConstantInt.
+  static std::pair<ConstantInt *, unsigned>
+  isBinOpWithConstantInt(Instruction *I) {
```
This is static.

HanKuanChen committed:
do no use magic number

HanKuanChen committed:
add const to isSame

HanKuanChen committed:
Merge remote-tracking branch 'upstream/main' into perf/slp-interchangeable-instruction-2

HanKuanChen committed:
fix merge conflict

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,285 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+  // For example, VL [x + 0, y * 1] can be converted to [x << 0, y << 0], but
+  // 'shl' does not exist in VL. In the end, we convert VL to [x * 1, y * 1].
+  // SeenBefore is used to know what operations have been seen before.
+  mutable MaskType SeenBefore = 0;
```
Avoid using mutable

HanKuanChen replied:
Then isSame cannot be const.

alexey-bataev replied:
Then isSame is wrong nname and it should not be const

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,285 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+  using InterchangeableInstruction::InterchangeableInstruction;
+  bool isSame(Instruction *I) const override {
+    unsigned Opcode = I->getOpcode();
+    if (!binary_search(SupportedOp, Opcode))
```
Add an assertion that SupportedOp is sorted

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,285 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+  }
+  unsigned getOpcode() const override {
+    MaskType Candidate = Mask & SeenBefore;
+    if (Candidate & SHL_BIT)
+      return Instruction::Shl;
```
You always have a priority here, which one is better to choose. No need to combine all possible alternatives using bitmasks, you can safely return single candidate

alexey-bataev replied:
IF the mask contains shl, it is always first candidate. If the mask contains ashr, it is always the second candidate. Why do you need to keep all candidates (in mask), if you have strict order, which one should be chosen?

HanKuanChen replied:
I don't get it. VL `[x << 0, y * 3]` cannot use Shl but can use mul to be `[x * 1, y * 3]`.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,6 +810,285 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+static bool getInterchangeableInstruction(
+    SmallVector<std::unique_ptr<InterchangeableInstruction>> &Candidate,
+    Instruction *I) {
+  auto Iter = std::stable_partition(
```
```suggestion
  auto *Iter = std::stable_partition(
```

HanKuanChen committed:
merge InterchangeableInstruction and InterchangeableBinOp

rename isSame to add
rename tryAnd to trySet
make Mask support MainOp_BIT

HanKuanChen committed:
reduce opcodeToMask(Opcode) usage

HanKuanChen committed:
use assert instead of llvm_unreachable

HanKuanChen committed:
rename mask and use LLVM_MARK_AS_BITMASK_ENUM

HanKuanChen committed:
support more pattern for add and sub

HanKuanChen committed:
add comment

HanKuanChen commented:
This is the compilation time result. https://llvm-compile-time-tracker.com/compare.php?from=a12ca57c1cb070be8e0048004c6b4e820029b6ee&to=b8c2e9e38d1e32cd2f23c928a50d83574b86c58f&stat=instructions:u

alexey-bataev commented:
> This is the compilation time result. https://llvm-compile-time-tracker.com/compare.php?from=a12ca57c1cb070be8e0048004c6b4e820029b6ee&to=b8c2e9e38d1e32cd2f23c928a50d83574b86c58f&stat=instructions:u

shared_sha256.c still requires some investigation

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -836,8 +1103,7 @@ class InstructionsState {
   bool isOpcodeOrAlt(Instruction *I) const {
-    unsigned CheckedOpcode = I->getOpcode();
-    return getOpcode() == CheckedOpcode || getAltOpcode() == CheckedOpcode;
+    return isConvertible(I, MainOp, AltOp);
```
Would be good to stop representing transformed nodes as nodes with the alternate InstructionsState. Instead, it should keep info that all instructions are homogeneous and build the correct operands for this, based on the the analysis, implemented here

alexey-bataev replied:
I mean the InstructionsState of "transformed" nodes should not consider them as alternate nodes, but as a regular nodes. Instead, need to introduce and build operands and operand nodes, using "transformed" operands

HanKuanChen committed:
refactor InterchangeableBinOp and add getInterchangeableMask

HanKuanChen committed:
add InterchangeableBinOp inside InstructionsState
add InterchangeableBinOp::contain
add InterchangeableBinOp::getMainOpConverter
add InterchangeableBinOp::getAltOpConverter
add InterchangeableBinOp::hasOpConverter

HanKuanChen committed:
refactor convertTo

HanKuanChen committed:
rename OpValue to Inst

HanKuanChen committed:
Revert "refactor convertTo"

This reverts commit ddcd456f749f1c4b12f7b96aa7c52197c0ac201a.

HanKuanChen committed:
Revert "add InterchangeableBinOp inside InstructionsState"

This reverts commit ad7bec92d18a7bc908d6717bdb49b69843ab323c.

HanKuanChen commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -810,11 +810,249 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
   Instruction *AltOp = nullptr;
+  // Only BinaryOperator will activate this.
+  std::optional<InterchangeableBinOp> MainOpConverter;
+  std::optional<InterchangeableBinOp> AltOpConverter;
```
Revert ad7bec92d18a7bc908d6717bdb49b69843ab323c and ddcd456f749f1c4b12f7b96aa7c52197c0ac201a to remove the data members.

HanKuanChen committed:
reduce binary_search usage

HanKuanChen committed:
refactor isBinOpWithConstantInt

HanKuanChen committed:
refactor InterchangeableBinOp to BinOpSameOpcodeHelper

HanKuanChen committed:
remove static: functions are in anonymous namespace already

HanKuanChen committed:
Merge remote-tracking branch 'upstream/main' into perf/slp-interchangeable-instruction-2

HanKuanChen committed:
fix merge

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1035,16 +1035,15 @@ class InterchangeableBinOp {
   }
 };
 
-static std::optional<InterchangeableBinOp> isConvertible(Instruction *From,
```
Why static is removed?

HanKuanChen replied:
these functions are in anonymous namespace already

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -813,6 +813,302 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+/// converted to another instruction with same semantics. For example, x << 1 is
+/// equal to x * 2. x * 1 is equal to x | 0.
+class BinOpSameOpcodeHelper {
+  using MaskType = std::uint_fast16_t;
```
Why not just unsigned short?

HanKuanChen replied:
I care about the type is the fastest type for at least 16 bits. unsigned short only means it has at least 16 bits, but does not mean it is the fastest.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -813,6 +813,302 @@ static std::optional<unsigned> getExtractIndex(Instruction *E) {
+    return {nullptr, 0};
+  }
+  struct InterchangeableInfo {
+    Instruction *I;
```
```suggestion
    Instruction *I = nullptr;
```

alexey-bataev commented on llvm/test/Transforms/SLPVectorizer/alternate-opcode-sindle-bv.ll:
```diff
@@ -14,6 +14,28 @@ define <2 x i32> @test(i32 %arg) {
 ; CHECK-NEXT:    [[TMP1:%.*]] = insertelement <2 x i32> [[TMP0]], i32 [[MUL]], i32 1
 ; CHECK-NEXT:    ret <2 x i32> [[TMP1]]
```
CHECK lines should be removed

alexey-bataev commented:
What about compile time?

HanKuanChen commented:
> What about compile time?

https://llvm-compile-time-tracker.com/compare.php?from=d77ef140e66dda44a40ac2f316d3b8a42fb7bf07&to=4dfe0132270ecd71dcef0f9a9616104b72d2207e&stat=instructions:u

HanKuanChen committed:
Update llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

Co-authored-by: Alexey Bataev <a.bataev@gmx.com>

HanKuanChen committed:
remove CHECK

HanKuanChen committed:
add allSameOpcode to fix llvm/test/Transforms/SLPVectorizer/X86/non-power-2-num-elems-reused.ll

HanKuanChen committed:
interchangeable instruction should apply to getOperandInfo

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2489,42 +2806,45 @@ class BoUpSLP {
+        Ops.resize(NumLanes);
+      for (unsigned Lane : seq<unsigned>(NumLanes)) {
+        assert((isa<Instruction>(VL[Lane]) || isa<PoisonValue>(VL[Lane])) &&
+               "Expected instruction or poison value");
+        // Our tree has just 3 nodes: the root and two operands.
+        // It is therefore trivial to get the APO. We only need to check the
+        // opcode of VL[Lane] and whether the operand at OpIdx is the LHS or RHS
+        // operand. The LHS operand of both add and sub is never attached to an
+        // inversese operation in the linearized form, therefore its APO is
+        // false. The RHS is true only if VL[Lane] is an inverse operation.
+
+        // Since operand reordering is performed on groups of commutative
+        // operations or alternating sequences (e.g., +, -), we can safely tell
+        // the inverse operations by checking commutativity.
+        if (isa<PoisonValue>(VL[Lane])) {
+          for (unsigned OpIdx : seq<unsigned>(NumOperands)) {
+            if (OpIdx == 0) {
+              if (auto *EI = dyn_cast<ExtractElementInst>(MainOp)) {
                 OpsVec[OpIdx][Lane] = {EI->getVectorOperand(), true, false};
                 continue;
               }
-            } else if (auto *EV = dyn_cast<ExtractValueInst>(MainOp)) {
-              if (OpIdx == 0) {
+              if (auto *EV = dyn_cast<ExtractValueInst>(MainOp)) {
                 OpsVec[OpIdx][Lane] = {EV->getAggregateOperand(), true, false};
                 continue;
               }
             }
             OpsVec[OpIdx][Lane] = {
                 PoisonValue::get(MainOp->getOperand(OpIdx)->getType()), true,
                 false};
-            continue;
           }
-          bool IsInverseOperation = !isCommutative(cast<Instruction>(VL[Lane]));
+          continue;
+        }
```
Can you do it in a separate NFC patch?

HanKuanChen replied:
https://github.com/llvm/llvm-project/pull/132152

HanKuanChen committed:
Merge remote-tracking branch 'upstream/main' into perf/slp-interchangeable-instruction-2

HanKuanChen committed:
fix merge conflict

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -813,6 +835,302 @@ static std::optional<unsigned> getExtractIndex(const Instruction *E) {
+///
+/// Example of unsupported opcode is SDIV that can potentially cause UB if the
+/// "shuffled out" lane would result in division by zero.
+bool isValidForAlternation(unsigned Opcode) {
+  if (Instruction::isIntDivRem(Opcode))
+    return false;
+
+  return true;
+}
```
```suggestion
bool isValidForAlternation(unsigned Opcode) {
  return !Instruction::isIntDivRem(Opcode);
}
```

HanKuanChen replied:
how about using `Instruction::isIntDivRem` directly?

alexey-bataev replied:
I think better to keep this function, just in case

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -813,6 +835,302 @@ static std::optional<unsigned> getExtractIndex(const Instruction *E) {
+        return Instruction::Xor;
+      llvm_unreachable("Cannot find interchangeable instruction.");
+    }
+    SmallVector<Value *> getOperand(Instruction *To) const {
```
Do you really need to return SmallVector here or just a pair is enough?

HanKuanChen replied:
pair is inconvenient. See the following code
```
        auto [SelectedOp, Ops] =
            convertTo(cast<Instruction>(VL[Lane]), MainOp, S.getAltOp());
        bool IsInverseOperation = !isCommutative(SelectedOp);
        for (unsigned OpIdx = 0; OpIdx != NumOperands; ++OpIdx) {
          bool APO = (OpIdx == 0) ? false : IsInverseOperation;
          OpsVec[OpIdx][Lane] = {Ops[OpIdx], APO, false};
        }
```
Ops can be accessed by index. pair cannot.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -813,6 +835,302 @@ static std::optional<unsigned> getExtractIndex(const Instruction *E) {
+           ::isValidForAlternation(I->getOpcode());
+  }
+  bool initializeAltOp(Instruction *I) {
+    if (!AltOp.I) {
+      if (!isValidForAlternation(I))
+        return false;
+      AltOp.I = I;
+    }
+    return true;
```
```suggestion
    if (AltOp.I)
      return true;
    if (!isValidForAlternation(I))
      return false;
    AltOp.I = I;
    return true;
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -813,6 +835,302 @@ static std::optional<unsigned> getExtractIndex(const Instruction *E) {
+    assert(is_sorted(SupportedOp) && "SupportedOp is not sorted.");
+  }
+  bool add(Instruction *I) {
+    assert(isa<BinaryOperator>(I));
```
Add a message

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -12369,14 +12689,22 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
+      // operand might differ from what is retrieved from UniqueValues[Idx].
+      Value *Op1 = E->getOperand(0)[Idx];
+      Value *Op2;
+      SmallVector<const Value *, 2> Operands({Op1});
```
```suggestion
      SmallVector<const Value *, 2> Operands(1, Op1);
```

HanKuanChen committed:
apply comment

alexey-bataev commented:
Rebase?

HanKuanChen committed:
Merge remote-tracking branch 'upstream/main' into perf/slp-interchangeable-instruction-2

HanKuanChen committed:
fix conflict

HanKuanChen committed:
apply comment

alexey-bataev reviewed (approved):
LG

rupprecht commented:
This commit causes a crash when building highway tests, e.g. https://github.com/google/highway/blob/master/hwy/tests/arithmetic_test.cc

Some tests (e.g. [unroller_test.cc](https://github.com/google/highway/blob/master/hwy/tests/unroller_test.cc)) fail with an assertion:
```
assert.h assertion failed at llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:18508 in void llvm::slpvectorizer::BoUpSLP::BlockScheduling::cancelScheduling(ArrayRef<Value *>, Value *): Bundle->isSchedulingEntity() && (Bundle->isPartOfBundle() || needToScheduleSingleInstruction(VL)) && "tried to unbundle something which is not a bundle"
*** Check failure stack trace: ***
    @     0x555d3e1577b0  llvm::slpvectorizer::BoUpSLP::BlockScheduling::cancelScheduling()
    @     0x555d3e11d390  llvm::slpvectorizer::BoUpSLP::BlockScheduling::tryScheduleBundle()
    @     0x555d3e1107f8  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
    @     0x555d3e11259f  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
    @     0x555d3e111620  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
    @     0x555d3e167823  llvm::SLPVectorizerPass::tryToVectorizeList()
    @     0x555d3e168afa  llvm::SLPVectorizerPass::tryToVectorize()
    @     0x555d3e16ba81  llvm::SLPVectorizerPass::vectorizeRootInstruction()
    @     0x555d3e16146f  llvm::SLPVectorizerPass::vectorizeChainsInBlock()
    @     0x555d3e15e54e  llvm::SLPVectorizerPass::runImpl()
    @     0x555d3e15dc5c  llvm::SLPVectorizerPass::run()
...
3.	Running pass "function<eager-inv>(float2int,lower-constant-intrinsics,chr,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O3>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "hwy/contrib/unroller/unroller_test.cc"
```

Other tests (e.g. [arithmetic_test.cc](https://github.com/google/highway/blob/master/hwy/tests/arithmetic_test.cc)) also crash with an identical stack, but with no assertion.

HanKuanChen commented:
> This commit causes a crash when building highway tests, e.g. https://github.com/google/highway/blob/master/hwy/tests/arithmetic_test.cc
> 
> Some tests (e.g. [unroller_test.cc](https://github.com/google/highway/blob/master/hwy/tests/unroller_test.cc)) fail with an assertion:
> 
> ```
> assert.h assertion failed at llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:18508 in void llvm::slpvectorizer::BoUpSLP::BlockScheduling::cancelScheduling(ArrayRef<Value *>, Value *): Bundle->isSchedulingEntity() && (Bundle->isPartOfBundle() || needToScheduleSingleInstruction(VL)) && "tried to unbundle something which is not a bundle"
> *** Check failure stack trace: ***
>     @     0x555d3e1577b0  llvm::slpvectorizer::BoUpSLP::BlockScheduling::cancelScheduling()
>     @     0x555d3e11d390  llvm::slpvectorizer::BoUpSLP::BlockScheduling::tryScheduleBundle()
>     @     0x555d3e1107f8  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
>     @     0x555d3e11259f  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
>     @     0x555d3e111620  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
>     @     0x555d3e167823  llvm::SLPVectorizerPass::tryToVectorizeList()
>     @     0x555d3e168afa  llvm::SLPVectorizerPass::tryToVectorize()
>     @     0x555d3e16ba81  llvm::SLPVectorizerPass::vectorizeRootInstruction()
>     @     0x555d3e16146f  llvm::SLPVectorizerPass::vectorizeChainsInBlock()
>     @     0x555d3e15e54e  llvm::SLPVectorizerPass::runImpl()
>     @     0x555d3e15dc5c  llvm::SLPVectorizerPass::run()
> ...
> 3.	Running pass "function<eager-inv>(float2int,lower-constant-intrinsics,chr,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O3>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "hwy/contrib/unroller/unroller_test.cc"
> ```
> 
> Other tests (e.g. [arithmetic_test.cc](https://github.com/google/highway/blob/master/hwy/tests/arithmetic_test.cc)) also crash with an identical stack, but with no assertion.

Any build command?

rupprecht commented:
> > This commit causes a crash when building highway tests, e.g. https://github.com/google/highway/blob/master/hwy/tests/arithmetic_test.cc
> > Some tests (e.g. [unroller_test.cc](https://github.com/google/highway/blob/master/hwy/tests/unroller_test.cc)) fail with an assertion:
> > ```
> > assert.h assertion failed at llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:18508 in void llvm::slpvectorizer::BoUpSLP::BlockScheduling::cancelScheduling(ArrayRef<Value *>, Value *): Bundle->isSchedulingEntity() && (Bundle->isPartOfBundle() || needToScheduleSingleInstruction(VL)) && "tried to unbundle something which is not a bundle"
> > *** Check failure stack trace: ***
> >     @     0x555d3e1577b0  llvm::slpvectorizer::BoUpSLP::BlockScheduling::cancelScheduling()
> >     @     0x555d3e11d390  llvm::slpvectorizer::BoUpSLP::BlockScheduling::tryScheduleBundle()
> >     @     0x555d3e1107f8  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
> >     @     0x555d3e11259f  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
> >     @     0x555d3e111620  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
> >     @     0x555d3e167823  llvm::SLPVectorizerPass::tryToVectorizeList()
> >     @     0x555d3e168afa  llvm::SLPVectorizerPass::tryToVectorize()
> >     @     0x555d3e16ba81  llvm::SLPVectorizerPass::vectorizeRootInstruction()
> >     @     0x555d3e16146f  llvm::SLPVectorizerPass::vectorizeChainsInBlock()
> >     @     0x555d3e15e54e  llvm::SLPVectorizerPass::runImpl()
> >     @     0x555d3e15dc5c  llvm::SLPVectorizerPass::run()
> > ...
> > 3.	Running pass "function<eager-inv>(float2int,lower-constant-intrinsics,chr,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O3>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "hwy/contrib/unroller/unroller_test.cc"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Other tests (e.g. [arithmetic_test.cc](https://github.com/google/highway/blob/master/hwy/tests/arithmetic_test.cc)) also crash with an identical stack, but with no assertion.
> 
> Any build command?

Hopefully better: a direct IR repro (this one is reduced from [zstd_opt.c](https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_opt.c)):

```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

%struct.ZSTD_optimal_t = type { i32, i32, i32, i32, [3 x i32] }

define i64 @ZSTD_compressBlock_opt_generic(ptr %ms, i32 %0, i32 %lastStretch.sroa.3.0.copyload) {
entry:
  %1 = load ptr, ptr %ms, align 8
  %sub506 = sub i32 %0, %lastStretch.sroa.3.0.copyload
  %add537 = add i32 %sub506, 1
  %idxprom556 = zext i32 %add537 to i64
  %arrayidx5573 = getelementptr %struct.ZSTD_optimal_t, ptr %1, i64 %idxprom556
  store i32 0, ptr %arrayidx5573, align 4
  %idxprom563.phi.trans.insert = zext i32 %sub506 to i64
  %arrayidx564.phi.trans.insert = getelementptr %struct.ZSTD_optimal_t, ptr %1, i64 %idxprom563.phi.trans.insert
  %.pre = load i8, ptr %arrayidx564.phi.trans.insert, align 1
  br label %while.body562

while.body562:                                    ; preds = %while.body562, %entry
  store i8 %.pre, ptr %ms, align 1
  br label %while.body562
}
```

Repros w/ `opt -S --passes=slp-vectorizer`

HanKuanChen commented:
> > > This commit causes a crash when building highway tests, e.g. https://github.com/google/highway/blob/master/hwy/tests/arithmetic_test.cc
> > > Some tests (e.g. [unroller_test.cc](https://github.com/google/highway/blob/master/hwy/tests/unroller_test.cc)) fail with an assertion:
> > > ```
> > > assert.h assertion failed at llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:18508 in void llvm::slpvectorizer::BoUpSLP::BlockScheduling::cancelScheduling(ArrayRef<Value *>, Value *): Bundle->isSchedulingEntity() && (Bundle->isPartOfBundle() || needToScheduleSingleInstruction(VL)) && "tried to unbundle something which is not a bundle"
> > > *** Check failure stack trace: ***
> > >     @     0x555d3e1577b0  llvm::slpvectorizer::BoUpSLP::BlockScheduling::cancelScheduling()
> > >     @     0x555d3e11d390  llvm::slpvectorizer::BoUpSLP::BlockScheduling::tryScheduleBundle()
> > >     @     0x555d3e1107f8  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
> > >     @     0x555d3e11259f  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
> > >     @     0x555d3e111620  llvm::slpvectorizer::BoUpSLP::buildTree_rec()
> > >     @     0x555d3e167823  llvm::SLPVectorizerPass::tryToVectorizeList()
> > >     @     0x555d3e168afa  llvm::SLPVectorizerPass::tryToVectorize()
> > >     @     0x555d3e16ba81  llvm::SLPVectorizerPass::vectorizeRootInstruction()
> > >     @     0x555d3e16146f  llvm::SLPVectorizerPass::vectorizeChainsInBlock()
> > >     @     0x555d3e15e54e  llvm::SLPVectorizerPass::runImpl()
> > >     @     0x555d3e15dc5c  llvm::SLPVectorizerPass::run()
> > > ...
> > > 3.	Running pass "function<eager-inv>(float2int,lower-constant-intrinsics,chr,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O3>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "hwy/contrib/unroller/unroller_test.cc"
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Other tests (e.g. [arithmetic_test.cc](https://github.com/google/highway/blob/master/hwy/tests/arithmetic_test.cc)) also crash with an identical stack, but with no assertion.
> > 
> > 
> > Any build command?
> 
> Hopefully better: a direct IR repro (this one is reduced from [zstd_opt.c](https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_opt.c)):
> 
> ```llvm
> target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
> target triple = "x86_64-unknown-linux-gnu"
> 
> %struct.ZSTD_optimal_t = type { i32, i32, i32, i32, [3 x i32] }
> 
> define i64 @ZSTD_compressBlock_opt_generic(ptr %ms, i32 %0, i32 %lastStretch.sroa.3.0.copyload) {
> entry:
>   %1 = load ptr, ptr %ms, align 8
>   %sub506 = sub i32 %0, %lastStretch.sroa.3.0.copyload
>   %add537 = add i32 %sub506, 1
>   %idxprom556 = zext i32 %add537 to i64
>   %arrayidx5573 = getelementptr %struct.ZSTD_optimal_t, ptr %1, i64 %idxprom556
>   store i32 0, ptr %arrayidx5573, align 4
>   %idxprom563.phi.trans.insert = zext i32 %sub506 to i64
>   %arrayidx564.phi.trans.insert = getelementptr %struct.ZSTD_optimal_t, ptr %1, i64 %idxprom563.phi.trans.insert
>   %.pre = load i8, ptr %arrayidx564.phi.trans.insert, align 1
>   br label %while.body562
> 
> while.body562:                                    ; preds = %while.body562, %entry
>   store i8 %.pre, ptr %ms, align 1
>   br label %while.body562
> }
> ```
> 
> Repros w/ `opt -S --passes=slp-vectorizer`

Cannot reproduce the error. My SHA-1 is 71a0cfd93263552ddc0bfd2ea7b0abe9a578f87e. Do you miss anything?

mstorsjo commented:
I also ran into failed asserts since this change, so I pushed a revert.

The issue is reproducible with the following reduced input:
```c
void make_filters_from_proto(int *filter[][2], int bands) {
  int c, q, n;
  for (;; q++) {
    n = 0;
    for (; n < 7; n++) {
      int theta = (q * (n - 6) + (n >> 1) - 3) % bands;
      if (theta)
        c = theta;
      filter[q][n][0] = c;
    }
  }
}
```

Compiled e.g. like this:
```
$ clang -target x86_64-linux-gnu -c repro.c -O3
clang: ../lib/Transforms/Vectorize/SLPVectorizer.cpp:989: llvm::SmallVector<llvm::Value*> {anonymous}::BinOpSameOpcodeHelper::InterchangeableInfo::getOperand(llvm::Instruction*) const: Assertion `FromCIValue.isZero() && "Cannot convert the instruction."' failed.
```

The issue also reproduces by compiling ffmpeg with e.g. llvm-test-suite, according to the instructions at https://github.com/llvm/llvm-test-suite/blob/main/External/ffmpeg/README.md.

glandium commented:
I'm also getting crashes when building libjpeg-turbo or libavcodec:
1.	<eof> parser at end of file
2.	Optimizer
3.	Running pass "function<eager-inv>(float2int,lower-constant-intrinsics,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O2>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "jidctint-151379.c"
4.	Running pass "slp-vectorizer" on function "jpeg_idct_12x12"
 #0 0x00007fbe6e0f1058 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/tmp/llvm/obj/bin/../lib/libLLVM.so.21.0git+0xea0058)
 #1 0x00007fbe6e0eeaae llvm::sys::RunSignalHandlers() (/tmp/llvm/obj/bin/../lib/libLLVM.so.21.0git+0xe9daae)
 #2 0x00007fbe6e0f1701 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x00007fbe6cd5c050 (/lib/x86_64-linux-gnu/libc.so.6+0x3c050)
 #4 0x00007fbe6fbb2393 (anonymous namespace)::BinOpSameOpcodeHelper::InterchangeableInfo::getOperand(llvm::Instruction*) const SLPVectorizer.cpp:0:0
 #5 0x00007fbe6fb4604b llvm::slpvectorizer::BoUpSLP::VLOperands::VLOperands(llvm::ArrayRef<llvm::Value*>, (anonymous namespace)::InstructionsState const&, llvm::slpvectorizer::BoUpSLP const&) SLPVectorizer.cpp:0:0
 #6 0x00007fbe6fb43630 llvm::slpvectorizer::BoUpSLP::TreeEntry::setOperand(llvm::slpvectorizer::BoUpSLP const&, bool) (/tmp/llvm/obj/bin/../lib/libLLVM.so.21.0git+0x28f2630)
 #7 0x00007fbe6fb346a8 llvm::slpvectorizer::BoUpSLP::buildTree_rec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/tmp/llvm/obj/bin/../lib/libLLVM.so.21.0git+0x28e36a8)
 #8 0x00007fbe6fb347e9 llvm::slpvectorizer::BoUpSLP::buildTree_rec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/tmp/llvm/obj/bin/../lib/libLLVM.so.21.0git+0x28e37e9)
 #9 0x00007fbe6fb3470d llvm::slpvectorizer::BoUpSLP::buildTree_rec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/tmp/llvm/obj/bin/../lib/libLLVM.so.21.0git+0x28e370d)
#10 0x00007fbe6fb347e9 llvm::slpvectorizer::BoUpSLP::buildTree_rec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/tmp/llvm/obj/bin/../lib/libLLVM.so.21.0git+0x28e37e9)
#11 0x00007fbe6fb347e9 llvm::slpvectorizer::BoUpSLP::buildTree_rec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/tmp/llvm/obj/bin/../lib/libLLVM.so.21.0git+0x28e37e9)

hiraditya commented:
Please revert this patch and put both patches for review.

HanKuanChen commented:
> Thanks @mstorsjo for reverting!
> 
> > Cannot reproduce the error. My SHA-1 is [71a0cfd](https://github.com/llvm/llvm-project/commit/71a0cfd93263552ddc0bfd2ea7b0abe9a578f87e). Do you miss anything?
> 
> Very certain it's this commit. Did you enable asserts in your local build? Hopefully the other repros can increase the chance of reproducability for you.

The error you encountered is different from the others. I would appreciate it if you could provide the steps to reproduce it. Here is my cmake command
```
cmake -DCMAKE_C_COMPILER=clang \
      -DCMAKE_CXX_COMPILER=clang++ \
      -DCMAKE_BUILD_TYPE=Debug \
      -DLLVM_APPEND_VC_REV=OFF \
      -DLLVM_ENABLE_ASSERTIONS=ON \
      -DLLVM_ENABLE_PROJECTS="clang;lld" \
      -DLLVM_ENABLE_RUNTIMES=all \
      -DLLVM_USE_LINKER=lld \
      -G Ninja ../llvm
```

glandium commented:
> I also ran into failed asserts since this change, so I pushed a revert.

Fun fact: the revert didn't fix the libavcodec crash for me, but fixed the libjpeg-turbo one.

glandium commented:
Actually, the revert fixed it, but this relanded in #132887 with a version that still crashes on libavcodec but not libjpeg-turbo.

HanKuanChen commented:
> Actually, the revert fixed it, but this relanded in #132887 with a version that still crashes on libavcodec but not libjpeg-turbo.

Can you give me the test?

glandium commented:
As reduced by cvise:
```
#define __intN_t(N, MODE) typedef int int##N##_t __attribute__((__mode__(MODE)))
__intN_t(16, __HI__);
#define itxfm_wrapper(type_a, type_b, sz, bits, has_dconly)                    \
  void type_a_type_b_szxsz_add_c(int16_t *_block) {                            \
    int i;                                                                     \
    int16_t *block = _block, tmp[sz];                                          \
    for (i = 0; i < sz; i++)                                                   \
      type_a##sz##_1d(block, sz, tmp + i * sz, 0);                             \
    type_b##sz##_1d(tmp);                                                      \
  }
#define itxfm_wrap(sz, bits) itxfm_wrapper(iadst, idct, sz, , )
#define IN(x) in[x]
int iadst4_1d_t1, iadst4_1d_t2, iadst4_1d_t3;
void idct4_1d();
void iadst4_1d(short *in, long, short *out, int) {
  iadst4_1d_t2 = iadst4_1d_t3 = IN(1);
  out[0] = out[1] = iadst4_1d_t1 + iadst4_1d_t3 >> 4;
  out[2] = iadst4_1d_t2 + 3 >> 4;
  out[3] = -iadst4_1d_t3 >> 4;
}
itxfm_wrap(4, )
```
Compile with these flags: `clang -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -relaxed-aliasing -target-cpu x86-64 -tune-cpu generic -O2 -vectorize-slp -x c `

mstorsjo commented:
I'm also still hitting crashes after the reland, also in libavcodec (but I hit it in another source file first, so I've reduced that one):
```c
short *transform_4x4_luma_12_coeffs;
int transform_4x4_luma_12_i;
void transform_4x4_luma_12() {
  transform_4x4_luma_12_i = 0;
  for (; transform_4x4_luma_12_i < 4; transform_4x4_luma_12_i++) {
    int c2 = -transform_4x4_luma_12_coeffs[3];
    transform_4x4_luma_12_coeffs[2] = transform_4x4_luma_12_coeffs[1] + 7 >> 7;
    transform_4x4_luma_12_coeffs[0] = transform_4x4_luma_12_coeffs[1] =
        transform_4x4_luma_12_coeffs[3] + transform_4x4_luma_12_coeffs[1] >> 7;
    transform_4x4_luma_12_coeffs[3] = c2 >> 7;
    transform_4x4_luma_12_coeffs += 4;
  }
}
```

Compiled with:
```
$ clang -target x86_64-linux-gnu -c repro.c -O3
```

I can push a revert in a little while.

And before relanding, at this point it would probably be good to test building all of ffmpeg with the suggested patch.
